packages/tpetra/core/src/Tpetra_Details_unpackCrsMatrix.hpp

```diff
@@ -70,270 +72,345 @@ class Distributor;
+
+  typedef UnpackCrsMatrixError<LO> value_type;
+
+  static_assert (std::is_same<LO, typename LocalMatrixType::ordinal_type>::value,
+                 "LocalMapType::local_ordinal_type and "
+                 "LocalMatrixType::ordinal_type must be the same.");
+
+  num_packets_per_lid_view_type num_packets_per_lid_;
+  offsets_view_type offsets_;
+  imports_view_type imports_;
+  import_lids_view_type import_lids_;
+  LocalMatrixType local_matrix_;
+  LocalMapType local_col_map_;
+  Tpetra::CombineMode combine_mode_;
+  bool atomic_;
+
+  UnpackCrsMatrixAndCombineFunctor(num_packets_per_lid_view_type num_packets_per_lid,
+      offsets_view_type offsets, imports_view_type imports,
+      import_lids_view_type import_lids,
+      local_matrix_type local_matrix, local_map_type local_col_map,
+      Tpetra::CombineMode combine_mode, bool atomic) :
+    num_packets_per_lid_(num_packets_per_lid), offsets_(offsets),
+    imports_(imports), import_lids_(import_lids), local_matrix_(local_matrix),
+    local_col_map_(local_col_map), combine_mode_(combine_mode), atomic_(atomic)
+  {}
+
+  KOKKOS_INLINE_FUNCTION void init(value_type& dst) const
+  {
+    dst.bad_index = Tpetra::Details::OrdinalTraits<LO>::invalid();
+    dst.num_ent = 0;
+    dst.offset = 0;
+    dst.expected_num_bytes = 0;
+    dst.num_bytes = 0;
+    dst.buf_size = 0;
+    dst.wrong_num_bytes_error = false;
+    dst.out_of_bounds_error = false;
   }
-  else {
-    TEUCHOS_TEST_FOR_EXCEPTION(
-        true, std::logic_error, "Invalid combine mode; should never get "
-        "here!  Please report this bug to the Tpetra developers.");
+
+  KOKKOS_INLINE_FUNCTION void
+  join (volatile value_type& dst, const volatile value_type& src) const
+  {
+    // The dst object should reflect the first bad index and all other
+    // associated error codes and data.  Thus, we need only check if the src
+    // object shows an error and if its associated `bad_index` is less than the
+    // dst.bad_index (if dst shows errors).
+    LO invalid = Tpetra::Details::OrdinalTraits<LO>::invalid();
+    if (src.bad_index != invalid) {
+      // An error in the src, check if whether:
+      //   1. The dst shows errors
+      //   2. If dst does show errors, if src bad_index is less than its bad
+      //      index
+      if (dst.bad_index == invalid || src.bad_index < dst.bad_index) {
+        dst.bad_index = src.bad_index;
+        dst.num_ent = src.num_ent;
+        dst.offset = src.offset;
+        dst.expected_num_bytes = src.expected_num_bytes;
+        dst.num_bytes = src.num_bytes;
+        dst.buf_size = src.buf_size;
+        dst.wrong_num_bytes_error = src.wrong_num_bytes_error;
+        dst.out_of_bounds_error = src.out_of_bounds_error;
+      }
+    }
   }
 
-  return numValid;
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const LO i, value_type& dst) const
+  {
+    const LO local_row = import_lids_[i];
+    const size_t num_bytes = num_packets_per_lid_(i);
```

Comment:
num_packets_per_lid_ is explicitly a HostSpace View.  You cannot access it on the GPU.  In general, whatever happens in a KOKKOS_INLINE_FUNCTION must be valid to run on the GPU.